amethyst_ui/src/widgets.rs

```diff
@@ -0,0 +1,208 @@
+    },
+    fmt::Display,
+    hash::Hash,
+    ops::Index,
+};
+
+/// A widget is an object that keeps track of all components and entities
+/// that make up an element of the user interface. Using the widget_components!
+/// macro, it's possible to generate methods that let you easily retrieve
+/// all components for a widget, and basically annotate which components the
+/// widget will definitely or maybe contain.
+/// Widgets are stored in their respective `Widgets` resource and referred to
+/// by their associated Id type. A widget will generally only contain fields
+/// for the entity Ids it consist of.
+pub trait Widget {}
+
+/// A WidgetId is the type by which the different widgets of a type will be
+/// differentiated when you create and retrieve them. Generally you'll want something
+/// here that can generate a random id with a low chance of collision, since
+/// auto generation will be used whenever you don't explicitly don't provide an
+/// id to widget builders.
+/// It's possible to use something like a bare enum as a WidgetId, but be aware
+/// that whenever you're not supplying a WidgetId, you'll probably overwrite an
+/// existing widget that had the same default id.
+pub trait WidgetId: Clone + PartialEq + Eq + Hash + Send + Sync + Display + 'static {
+    /// Generate a new widget id. This function can optionally be passed the last ID
+    /// that was generated, to make sequential ids possible.
+    fn generate(last: &Option<Self>) -> Self;
+}
+
+impl WidgetId for u32 {
+    fn generate(last: &Option<Self>) -> Self {
+        last.map(|last| last + 1).unwrap_or(0)
+    }
+}
+
+impl WidgetId for u64 {
+    fn generate(last: &Option<Self>) -> Self {
+        last.map(|last| last + 1).unwrap_or(0)
+    }
+}
+
+impl WidgetId for String {
+    fn generate(_: &Option<Self>) -> Self {
+        rand::thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(16)
+            .collect()
+    }
+}
+
+/// Widgets is an alias for a HashMap containing widgets mapped by their
+/// respective Id type. It's meant to be used as a resource for every widget type.
+#[derive(Derivative)]
+#[derivative(Default(bound = ""))]
+pub struct Widgets<T: Widget, I: WidgetId = u32> {
+    items: HashMap<I, T>,
+    last_key: Option<I>,
+}
+
+impl<T, I> Widgets<T, I>
+where
+    T: Widget,
+    I: WidgetId,
+{
+    // TODO(happens): Should we have methods here that safely add something,
+    // e.g. don't add in when you would overwrite something? Also, should we
+    // notify when something is overwritten?
+
+    /// Adds a widget to the map and returns the ID that was created
+    /// for it.
+    pub fn add(&mut self, widget: T) -> I {
+        let id = I::generate(&self.last_key);
+        self.items.insert(id.clone(), widget);
+        id
+    }
+
+    /// Adds a widget with a specified ID. This will overwrite any
+    /// existing widget of the same type with the same ID.
+    pub fn add_with_id(&mut self, id: I, widget: T) -> I {
```

Comment:
Since the ID is passed by the user, I'd return `Option<T>` here so you can check if it did overwrite the entry.

Reply:
So, your suggestion is that this returns `(I, Option<T>)` and that `Option<T>` will be `Some(widget)` containing the overwritten widget if it was already there? I feel like that leads to a weird pattern of users always feeling like they need to check the return value... Maybe we can add another method that does `add_or_replace` which can have your suggested signature?

Reply:
No, I think it should _only_ return `Option<T>`, since the user can clone the ID themselves.

Reply:
Ah, I see. To be honest I would prefer keeping the interface as-is, so that it stays the same as `add`, and also since this consumes the ID I think this would be easier to use. Would you be OK with my above suggestion of adding an explicit non-overwrite method?

Reply:
Yeah that makes sense. Hmm, but that doesn't cover the case where you do want to overwrite, but acquire ownership over the old widget, does it?

I'm also happy to leave this up to a future PR as written in the TODO comment.

Reply:
Nevermind, now that I think about it, doing it your way mimics the behavior of `HashMap` and is probably more familiar to most people. And there's no way to avoid a clone with my method if the user doesn't care about keeping the ID, so yours is more flexible.